setup.py

```diff
@@ -97,63 +70,79 @@ def run(self):
             self.build_extension(ext)
 
     def build_extension(self, ext):
-        self.compile(self.build_temp)
-
-        so_file = f"{ext.name}.so"
-        try:
-            os.remove(so_file)
-        except:
-            pass
-
-        if not osp.exists(so_file):
-            extension_file = osp.join(os.path.dirname(self.get_ext_fullpath(ext.name)), so_file)
-
-            print(f"Creating symlink from '{so_file}' to '{extension_file}'")
-            os.symlink(extension_file, so_file)
-
-    def compile(self, build_dir):
+        dest = os.path.abspath(self.get_ext_fullpath(ext.name))
+        print(f"Destination: {dest}")
+        install_args = [f"-DCMAKE_INSTALL_PREFIX={os.path.dirname(dest)}"]
+
+        cmake = shutil.which("cmake")  # Windows safe
+        if not cmake and platform.system() != "Windows":
+            print("Did not find cmake on the path. Will try to resolve from shell", file=sys.stderr)
+            cmake = subprocess.check_output(["bash", "-lc", "which cmake | tail -n 1"], universal_newlines=True).strip()
+
+        env_var = "ARCTIC_CMAKE_PRESET"  # From CMakePresets.json
+        preset = os.getenv(env_var, "*")
+        if preset == "skip":
+            return
+        search = f"cpp/out/{preset}-build"
+        candidates = glob.glob(search)
+        if not candidates:
+            common_args = [
+                "-DTEST=NO",
+                f"-DBUILD_PYTHON_VERSION={sys.version_info[0]}.{sys.version_info[1]}",
+                *install_args,
+            ]
+            cmake_check = subprocess.run([cmake, "-P", "cpp/CMake/CheckSupportsPreset.cmake"])
+            if cmake_check.returncode == 0:
+                self._configure_cmake_using_preset(cmake, common_args, preset, search)
+            else:
+                self._configure_for_legacy_image(common_args, "cpp/out/legacy-build")
+            candidates = glob.glob(search)
+
+        assert len(candidates) == 1, f"Specify {env_var} or use a single build directory. {search}={candidates}"
+        subprocess.check_call([cmake, "--build", candidates[0], "--target", ext.name])
+        subprocess.check_call(
+            [cmake, *install_args, "-DCOMPONENT=Python_Lib", "-P", candidates[0] + "/cmake_install.cmake"]
+        )
+
+        assert os.path.exists(dest), f"No output at {dest}, but we didn't get a bad return code from CMake?"
+
+    def _configure_cmake_using_preset(self, cmake, common_args, preset, search):
+        if preset == "*":
+            suffix = "-debug" if self.debug else "-release"
+            preset = ("windows-cl" if platform.system() == "Windows" else platform.system().lower()) + suffix
+        print(
+            f"Did not find build directory with '{search}'. Will configure and build using cmake preset {preset}",
+            file=sys.stderr,
+        )
+        subprocess.check_call([cmake, *common_args, "--preset", preset], cwd="cpp")
+
+    def _configure_for_legacy_image(self, common_args, build_dir):
```

Comment:
Probably cleaner to just update the cmake in the internal Docker image than support legacy CMAKE?

Reply:
Not wanting to spend another 3 weeks fixing compilation dependency issues :P